File: src/ParentConsole.cs
// Docked Console, a terminal docked to the edge of the Windows desktop.
// Copyright (C) 2026 Locke Werks
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see <https://www.gnu.org/licenses/>.
namespace DockedConsole;

/// <summary>
/// The dock is a GUI-subsystem executable so that running it never opens a console of its
/// own. That also means its control commands have nowhere to print. Borrowing the calling
/// shell's console gives --stop, --reclaim and --help somewhere to report.
/// </summary>
internal static class ParentConsole
{
    private static bool _attached;

    public static bool Attach()
    {
        if (_attached)
        {
            return true;
        }

        if (!Native.AttachConsole(Native.ATTACH_PARENT_PROCESS))
        {
            return false;
        }

        try
        {
            // The runtime bound stdout to a null writer at startup because there was no
            // console then. Rebind it now that there is one.
            var stdout = new StreamWriter(Console.OpenStandardOutput()) { AutoFlush = true };
            var stderr = new StreamWriter(Console.OpenStandardError()) { AutoFlush = true };
            Console.SetOut(stdout);
            Console.SetError(stderr);
            _attached = true;
        }
        catch (IOException)
        {
            return false;
        }

        return _attached;
    }

    public static void Detach()
    {
        if (!_attached)
        {
            return;
        }

        try
        {
            Console.Out.Flush();
            Console.Error.Flush();
        }
        catch (IOException)
        {
            // Nothing to do if the parent's console has already gone away.
        }

        Native.FreeConsole();
        _attached = false;
    }

    /// <summary>
    /// Prints to the calling shell if there is one and stays silent otherwise.
    ///
    /// Used by anything an installer hook can invoke. A hook has no console, and the
    /// dialog fallback in Report would block until the hook's timeout expires and fail
    /// the install with no one there to click OK.
    /// </summary>
    public static void Write(string message, bool error = false)
    {
        if (!Attach())
        {
            return;
        }

        if (error)
        {
            Console.Error.WriteLine(message);
        }
        else
        {
            Console.WriteLine(message);
        }
    }

    /// <summary>Prints to the calling shell, falling back to a dialog when there is none.</summary>
    public static void Report(string message, bool error = false)
    {
        if (Attach())
        {
            if (error)
            {
                Console.Error.WriteLine(message);
            }
            else
            {
                Console.WriteLine(message);
            }

            return;
        }

        MessageBox.Show(
            message,
            "Docked Console",
            MessageBoxButtons.OK,
            error ? MessageBoxIcon.Warning : MessageBoxIcon.Information);
    }
}

File: src/parent_console.cpp
// Docked Console, a terminal docked to the edge of the Windows desktop.
// Copyright (C) 2026 Locke Werks
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see <https://www.gnu.org/licenses/>.
#include "parent_console.h"

#include "win32.h"

#include <string>

namespace dock::console {
namespace {

bool g_attached = false;

bool Attach()
{
    if (g_attached) {
        return true;
    }
    if (!AttachConsole(ATTACH_PARENT_PROCESS)) {
        return false;
    }
    g_attached = true;
    return true;
}

/// The standard handle when the caller supplied one, otherwise the console we
/// just borrowed. Falling back to CONOUT$ matters because a GUI process started
/// from Explorer has no standard handles at all.
HANDLE StreamHandle(bool error)
{
    const HANDLE inherited = GetStdHandle(error ? STD_ERROR_HANDLE : STD_OUTPUT_HANDLE);
    if (inherited && inherited != INVALID_HANDLE_VALUE) {
        return inherited;
    }

    static Handle conout(CreateFileW(L"CONOUT$", GENERIC_WRITE,
                                     FILE_SHARE_READ | FILE_SHARE_WRITE, nullptr,
                                     OPEN_EXISTING, 0, nullptr));
    return conout.Get();
}

void Emit(std::wstring_view message, bool error)
{
    const HANDLE stream = StreamHandle(error);
    if (!stream || stream == INVALID_HANDLE_VALUE) {
        return;
    }

    std::wstring line(message);
    line.append(L"\r\n");

    // GetConsoleMode succeeds only for a real console. When the caller
    // redirected us into a file or a pipe, WriteConsoleW would fail, so the
    // bytes have to go out as UTF-8 through WriteFile instead.
    DWORD mode = 0;
    if (GetConsoleMode(stream, &mode)) {
        DWORD written = 0;
        WriteConsoleW(stream, line.data(), static_cast<DWORD>(line.size()),
                      &written, nullptr);
        return;
    }

    const std::string utf8 = Narrow(line);
    DWORD written = 0;
    WriteFile(stream, utf8.data(), static_cast<DWORD>(utf8.size()), &written, nullptr);
}

} // namespace

void Write(std::wstring_view message, bool error)
{
    if (!Attach()) {
        return;
    }
    Emit(message, error);
}

void Report(std::wstring_view message, bool error)
{
    if (Attach()) {
        Emit(message, error);
        return;
    }

    const std::wstring text(message);
    MessageBoxW(nullptr, text.c_str(), L"Docked Console",
                MB_OK | (error ? MB_ICONWARNING : MB_ICONINFORMATION));
}

void Detach()
{
    if (!g_attached) {
        return;
    }
    FreeConsole();
    g_attached = false;
}

} // namespace dock::console
